app/src/main/java/ru/ifmo/md/extratask1/ImageCacher.java

```diff
@@ -0,0 +1,49 @@
+package ru.ifmo.md.extratask1;
+
+import android.content.Context;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.widget.ImageView;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+public class ImageCacher {
+    private String cacheDir;
+
+    public ImageCacher(Context ctx) {
+        cacheDir = ctx.getCacheDir().getAbsolutePath();
+    }
+
+    public boolean isAvailable(String res) {
+        return getFile(res).exists();
+    }
+
+    public void set(String res, InputStream is) {
+        File file = getFile(res);
+        try {
+            FileOutputStream fos = new FileOutputStream(file);
+            int bufferSize = 4096;
+            byte[] buffer = new byte[bufferSize];
+            int len;
+            while ((len = is.read(buffer)) != -1) {
+                fos.write(buffer, 0, len);
+            }
+            fos.close();
+        } catch (IOException e) {
+            // should not occur here
+        }
+    }
+
+    public void putToImageView(ImageView imgView, String res) {
+        String filename = getFile(res).getAbsolutePath();
+        Bitmap bitmap = BitmapFactory.decodeFile(filename);
+        imgView.setImageBitmap(bitmap);
```

Comment:
As you have previously pointed out, doing IO & bitmap decoding operations in UI thread is a very bad idea. This is exactly what happens in this case, as putToImageView() is always called in UI thread.
We can fix this by making putToImageView() return Bitmap and set this bitmap to imgView in UI thread. I firstly implemented it that way, but then I changed it to the current version, because I thought transferring Bitmaps between UI/worker thread would introduce hard job for garbage collector as it would be harder to count links to these Bitmaps. However, changing putToImageView() to take ImageView rather than returning Bitmap didn't change anything in terms of UX, so apparently my assumption was wrong, but I forgot to fix putToImageView() back anyway.
I would gladly fix this now, when I know how bad it is for UI jobs and any other jobs to be in the same thread, but then I'd lose my precious 10 points for late commits. So I just wanted you to know that I know that there's bug here, lately realised, however.